File: bindings/kotlin/src/main/kotlin/org/equalitie/ouisync/File.kt
package org.equalitie.ouisync

class File private constructor(private val handle: Long, private val client: Client) {
    companion object {
        /**
         * Creates a new file at the given path in the repository.
         */
        suspend fun create(repo: Repository, path: String): File {
            val handle = repo.client.invoke(FileCreate(repo.handle, path)) as Long
            return File(handle, repo.client)
        }

        /**
         * Opens an existing file at the given path in the repository.
         */
        suspend fun open(repo: Repository, path: String): File {
            val handle = repo.client.invoke(FileOpen(repo.handle, path)) as Long
            return File(handle, repo.client)
        }

        /**
         * Removes file at the given path from the repository.
         */
        suspend fun remove(repo: Repository, path: String) {
            repo.client.invoke(FileRemove(repo.handle, path))
        }
    }

    suspend fun close() = client.invoke(FileClose(handle))

    suspend fun flush() = client.invoke(FileFlush(handle))

    /**
     * Length of this file in bytes
     */
    suspend fun length() = client.invoke(FileLen(handle)) as Long

    /**
     * Read the given amount of bytes from this file, starting at the given offset.
     */
    suspend fun read(offset: Long, length: Long) =
        client.invoke(FileRead(handle, offset, length)) as ByteArray

    /**
     * Write the content of the array to the file at the given offset.
     */
    suspend fun write(offset: Long, array: ByteArray) =
        client.invoke(FileWrite(handle, offset, array))

    /**
     * Truncate the file to the given length.
     */
    suspend fun truncate(length: Long) = client.invoke(FileTruncate(handle, length))

    /**
     * Sync progress of this file, that is, what part of this file (in bytes) is available locally.
     */
    suspend fun progress() = client.invoke(FileProgress(handle)) as Long
}

File: bindings/swift/Ouisync/Sources/File.swift
//
//  File.swift
//  OuisyncLib
//
//  Created by Radu Dan on 17.01.2025 and this generated comment was preserved because:
//
//  How many times throughout Xcode's history has `File.swift` ever been the intended name?

import Foundation
import MessagePack


public extension Repository {
    /** Opens an existing file from the current repository at `path`.
     *
     * Throws `OuisyncError` if `path` doesn't exist or is a directory. */
    func openFile(at path: String) async throws -> File {
        try await File(self, client.invoke("file_open", with: ["repository": handle,
                                                               "path": .string(path)]))
    }

    /** Creates a new file at `path`.
     *
     * Throws `OuisyncError` if `path` already exists of if the parent folder doesn't exist. */
    func createFile(at Path: String) async throws -> File {
        try await File(self, client.invoke("file_create", with: ["repository": handle,
                                                                 "path": .string(path)]))
    }

    /// Removes (deletes) the file at `path`.
    func removeFile(at path: String) async throws {
        try await client.invoke("file_remove", with: ["repository": handle,
                                                      "path": .string(path)])
    }
}


public class File {
    let repository: Repository
    let handle: MessagePackValue

    init(_ repo: Repository, _ handle: MessagePackValue) throws {
        _ = try handle.uint64Value.orThrow
        repository = repo
        self.handle = handle
    }

    /// Flush and close the handle once the file goes out of scope
    deinit {
        let client = repository.client, handle = handle
        Task { try await client.invoke("file_close", with: handle) }
    }
}


public extension File {
    /** Reads and returns at most `size` bytes from this file, starting at `offset`.
     *
     * Returns 0 bytes if `offset` is at or past the end of the file */
    func read(_ size: UInt64, fromOffset offset: UInt64) async throws -> Data {
        try await repository.client.invoke("file_read",
                                           with: ["file": handle,
                                                  "offset": .uint(offset),
                                                  "size": .uint(size)]).dataValue.orThrow
    }

    /// Writes `data` to this file, starting at `offset`
    func write(_ data: Data, toOffset offset: UInt64) async throws -> Data {
        try await repository.client.invoke("file_write",
                                           with: ["file": handle,
                                                  "offset": .uint(offset),
                                                  "data": .binary(data)]).dataValue.orThrow
    }

    /// Flushes any pending writes to persistent storage.
    func flush() async throws {
        try await repository.client.invoke("file_flush", with: handle)
    }

    /// Truncates the file to `len` bytes.
    func truncate(to len: UInt64 = 0) async throws {
        try await repository.client.invoke("file_truncate", with: ["file": handle,
                                                                   "len": .uint(len)])
    }

    var len: UInt64 { get async throws {
        try await repository.client.invoke("file_len", with: handle).uint64Value.orThrow
    } }

    var progress: UInt64 { get async throws {
        try await repository.client.invoke("file_progress", with: handle).uint64Value.orThrow
    } }
}
